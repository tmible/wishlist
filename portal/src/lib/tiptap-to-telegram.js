/** @module Перевод дерева текстового редактора в набор элементов разметки текста телеграма */

/** @typedef {import('@tiptap/core').JSONContent} JSONContent */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/**
 * @template ArrayType
 * @typedef
 * {ArrayType extends readonly (infer ElementType)[] ? ElementType : never}
 * ArrayElement<ArrayType extends readonly unknown[]>
 */
// eslint-disable-next-line no-secrets/no-secrets -- Просто длинное название
/**
 * Уплощёный узел дерева текстового редактора
 * @typedef {object} FlatChild
 * @property {string[]} marks Набор активных меток
 * //Будет работать только в TS
 * //@property {Pick<ArrayElement<Pick<JSONContent, 'marks'>>, 'type'>[]} marks Набор активных меток
 * @property {Pick<JSONContent, 'text'>} text Содержащийся в узле текст
 * @property {Pick<JSONContent, 'attrs'>} attrs Атрибуты для меток
 */

/**
 * Отображение названий меток текстового редактора в типы элементов
 * разметки текста телегарма Не упомянутые совпадают
 * @constant {Map<string, Pick<Entity, 'type'>>}
 */
const TIPTAP_TO_TELEGRAM_NAMES_MAP = new Map([
  [ 'strike', 'strikethrough' ],
  [ 'codeBlock', 'pre' ],
  [ 'link', 'url' ],
]);

/**
 * Рекурсивное уплощение дерева текстового редактора
 * @function flatContent
 * @param {JSONContent} content Текущий узел дерева
 * @param {string[]} [mixin] Примесь меток от родительских узлов дерева
 * @returns {FlatChild[]} Массив уплощённых потомков текущего узла
 */
const flatContent = (content, mixin = []) => content?.flatMap((child) => {
  if (child.type === 'text') {
    const url = child.marks?.find(({ type }) => type === 'text_link')?.attrs.href;
    return {
      marks: [ ...mixin, ...(child.marks?.map(({ type }) => type) ?? []) ],
      text: child.text,
      ...(url ? { attrs: { url } } : {}),
    };
  }

  if (child.type === 'codeBlock') {
    return {
      marks: [ child.type ],
      text: child.content[0].text ?? '',
      attrs: child.attrs,
    };
  }

  if (child.type === 'blockquote') {
    return flatContent(child.content, [ ...mixin, child.type ]);
  }

  const childContentFlat = flatContent(child.content, mixin) ?? [];
  if (child.content) {
    childContentFlat.at(-1).text += '\n';
  }
  return childContentFlat;
});

/**
 * Дополнительные проверки для приведения к соответствию формату Телеграма
 * @function preAndTextLinksMapper
 * @param {Entity} entity Элемент разметки текста
 * @returns {Entity} Обновлённый элемент разметки текста
 */
const preAndTextLinksMapper = (entity) => {
  if (entity.type === 'pre' && entity.language === null) {
    delete entity.language;
  }
  if (entity.type === 'text_link' && !entity.url.endsWith('/')) {
    entity.url += '/';
  }
  return entity;
};

/**
 * Перевод дерева текстового редактора в набор элементов разметки текста телеграма
 * 1. [Уплощение дерева]{@link flatContent};
 * 2. Перебор получившихся блоков:
 * 2.1. Создание элементов разметки текста на основе меток, закончившихся на начале текущего блока;
 * 2.2. Добавление к активным меткам меток, начавшихся в начале текущего блока;
 * 2.3. Добавление к тексту текста текущего блока;
 * 3. Удаление пробельных символов с конца текста;
 * 4. Создание элементов разметки текста на основе меток, закончившихся в конце текста;
 * @function tiptapToTelegram
 * @param {JSONContent} content Корень дерева текстового редактора
 * @returns {[ string, Entity[] ]} Текст и набор элементов его разметки
 */
export const tiptapToTelegram = (content) => {
  const contentFlat = flatContent(content);

  let text = '';
  const entities = [];
  let activeMarks = [];

  for (
    let i = 0, finishedMarks = [], [ block ] = contentFlat;
    i < contentFlat.length;
    ++i, text += block.text, block = contentFlat[i]
  ) {
    ({ activeMarks = [], finishedMarks = [] } = Object.groupBy(
      activeMarks,
      ({ type }) => (block.marks.includes(type) ? 'activeMarks' : 'finishedMarks'),
    ));

    entities.push(
      ...finishedMarks
        .map((finishedMark) => ({
          ...finishedMark,
          type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(finishedMark.type) ?? finishedMark.type,
          length: text.length - finishedMark.offset,
        }))
        .map((entity) => preAndTextLinksMapper(entity)),
    );

    activeMarks.push(
      ...block.marks
        .filter((mark) => !activeMarks.some(({ type }) => type === mark))
        .map((mark) => ({ ...block.attrs, offset: text.length, type: mark })),
    );
  }

  text = text.trimEnd();

  entities.push(
    ...activeMarks
      .map((activeMark) => ({
        ...activeMark,
        type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(activeMark.type) ?? activeMark.type,
        length: text.length - activeMark.offset,
      }))
      .map((entity) => preAndTextLinksMapper(entity)),
  );

  return [ text, entities ];
};
