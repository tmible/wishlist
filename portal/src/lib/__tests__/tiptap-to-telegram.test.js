import { describe, expect, it } from 'vitest';
import { tiptapToTelegram } from '../tiptap-to-telegram.js';

/* eslint-disable-next-line @stylistic/js/max-len --
   Длинная константа, читать не нужно, хранить так проще */
const content = JSON.parse('[{"type":"paragraph","content":[{"type":"text","marks":[{"type":"bold"}],"text":"ueyrghi"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"italic"}],"text":"erogij"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"underline"}],"text":"woeriugh"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"strike"}],"text":"woerifrjh"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"code"}],"text":"woeifj"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"spoiler"}],"text":"woeihjf"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"bold"}],"text":"woeifn "},{"type":"text","marks":[{"type":"bold"},{"type":"strike"}],"text":"woerifj"},{"type":"text","marks":[{"type":"bold"}],"text":" woefij"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"italic"}],"text":"woiefj "},{"type":"text","marks":[{"type":"italic"},{"type":"underline"}],"text":"owiejf"},{"type":"text","marks":[{"type":"underline"}],"text":" weiufn"}]},{"type":"codeBlock","attrs":{"language":null},"content":[{"type":"text","text":"oreigjnpoerkg"}]},{"type":"blockquote","content":[{"type":"paragraph","content":[{"type":"text","text":"lsrgnj"}]},{"type":"paragraph","content":[{"type":"text","text":"epork"},{"type":"text","marks":[{"type":"spoiler"}],"text":"g "},{"type":"text","marks":[{"type":"spoiler"},{"type":"strike"}],"text":"erkmg"},{"type":"text","marks":[{"type":"spoiler"}],"text":" wpoerg"},{"type":"text","text":" wpoef"}]}]},{"type":"paragraph","content":[{"type":"text","marks":[{"type":"mention","attrs":{"href":"https://t.me/wlkergn","target":"_blank","rel":"noopener noreferrer nofollow","class":"cursor-pointer"}}],"text":"@wlkergn"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"hashtag"}],"text":"#erkbn"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"cashtag"}],"text":"$KDF"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"bot_command"}],"text":"/ernm"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"email","attrs":{"href":"mailto:ev@jewpeorij.lnjf","target":"_blank","rel":"noopener noreferrer nofollow","class":"cursor-pointer"}}],"text":"ev@jewpeorij.lnjf"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"link","attrs":{"href":"https://example.com","target":"_blank","rel":"noopener noreferrer nofollow","class":"cursor-pointer"}}],"text":"example.com"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"phone_number","attrs":{"href":"tel:+12345678901","target":"_blank","rel":"noopener noreferrer nofollow","class":"cursor-pointer"}}],"text":"+12345678901"},{"type":"text","text":" "},{"type":"text","marks":[{"type":"text_link","attrs":{"href":"https://example.com","class":"cursor-pointer"}}],"text":"ldtkrhnj"}]}]');

describe('tiptapToTelegram', () => {
  it('should convert tiptap tree to string and telegram message entities', () => {
    expect(tiptapToTelegram(content)).toMatchSnapshot();
  });
});
