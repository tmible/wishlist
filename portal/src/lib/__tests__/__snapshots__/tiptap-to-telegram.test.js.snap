// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tiptapToTelegram > should convert tiptap tree to string and telegram message entities 1`] = `
[
  "ueyrghi erogij woeriugh woerifrjh woeifj woeihjf woeifn woerifj woefij woiefj owiejf weiufn
oreigjnpoerkglsrgnj
eporkg erkmg wpoerg wpoef
@wlkergn #erkbn $KDF /ernm ev@jewpeorij.lnjf example.com +12345678901 ldtkrhnj",
  [
    {
      "length": 7,
      "offset": 0,
      "type": "bold",
    },
    {
      "length": 6,
      "offset": 8,
      "type": "italic",
    },
    {
      "length": 8,
      "offset": 15,
      "type": "underline",
    },
    {
      "length": 9,
      "offset": 24,
      "type": "strikethrough",
    },
    {
      "length": 6,
      "offset": 34,
      "type": "code",
    },
    {
      "length": 7,
      "offset": 41,
      "type": "spoiler",
    },
    {
      "length": 7,
      "offset": 56,
      "type": "strikethrough",
    },
    {
      "length": 21,
      "offset": 49,
      "type": "bold",
    },
    {
      "length": 13,
      "offset": 71,
      "type": "italic",
    },
    {
      "length": 14,
      "offset": 78,
      "type": "underline",
    },
    {
      "length": 13,
      "offset": 92,
      "type": "pre",
    },
    {
      "length": 5,
      "offset": 119,
      "type": "strikethrough",
    },
    {
      "length": 14,
      "offset": 117,
      "type": "spoiler",
    },
    {
      "length": 33,
      "offset": 105,
      "type": "blockquote",
    },
    {
      "length": 8,
      "offset": 138,
      "type": "mention",
    },
    {
      "length": 6,
      "offset": 147,
      "type": "hashtag",
    },
    {
      "length": 4,
      "offset": 154,
      "type": "cashtag",
    },
    {
      "length": 5,
      "offset": 159,
      "type": "bot_command",
    },
    {
      "length": 17,
      "offset": 165,
      "type": "email",
    },
    {
      "length": 11,
      "offset": 183,
      "type": "url",
    },
    {
      "length": 12,
      "offset": 195,
      "type": "phone_number",
    },
    {
      "length": 8,
      "offset": 208,
      "type": "text_link",
      "url": "https://example.com/",
    },
  ],
]
`;
